### PR TITLE
new P2P info message

### LINKED ISSUE #3107: Info p2p message
An additional "info" (or other name) [P2P protocol message](https://github.com/aeternity/protocol/blob/master/sync/p2p_messages.md) proving more information about the node would be very helpful to monitor the network health and adoption rate.

As a minim it should return the node "version".
Additional fields would be architecture, OS, network_id, peers_count etc. similar to what we have in /status endpoint.

### PR DESCRIPTION
**Pull request by velzevur**
Fixes #3107 

The work on this PR had been supported by the Aeternity Foundation.

### CODE REVIEW

**Comment by velzevur** [comment]
Corresponding protocol PR will follow

File: apps/aecore/src/aec_peer_connection.erl [hunk 1]
```diff
@@ -1443,3 +1461,30 @@ validate_micro_signature(MicroHeader, _PrevHeader, PrevKeyHeader)
...
+handle_get_node_info(S) ->
+    case is_node_info_sharing_enabled() of
+        false -> S;
+        true ->
+            NodeVersion = integer_to_binary(?KEY_HEADER_INFO_LIMA_POINT_RELEASE),
+            OS = aeu_info:get_os(),
+            NodeInfo = #{ version => NodeVersion
+                        , os => OS
+                        , network_id => aec_governance:get_network_id()
+                        , verified_peers => 2
```
Comment (uwiger) [inline apps/aecore/src/aec_peer_connection.erl:1475]: ??

File: apps/aecore/src/aec_sync.erl [hunk 2]
```diff
@@ -1117,3 +1136,89 @@ pp_sync_task(ST = #sync_task{}) ->
...
+                  true -> none;
+                  false -> M 
+              end
+          end,
+      Aggr =
+          fun(Key) ->
+              Res =
+                  lists:foldl(
+                      fun(#{Key := NodeVsn}, Accum) ->
+                          maps:put(NodeVsn, maps:get(NodeVsn, Accum, 0) + 1, Accum)
```
Comment (uwiger) [inline apps/aecore/src/aec_sync.erl:1169]:
```suggestion
                          maps:update_with(NodeVsn, fun(X) -> X + 1 end, 0, Accum)
```

File: apps/aecore/src/aec_sync.erl [hunk 3]
```diff
@@ -1117,3 +1136,89 @@ pp_sync_task(ST = #sync_task{}) ->
...
+                      Result =
+                          receive
+                              {pmap_result_inner, R} -> R
+                          after Timeout -> {error, request_timeout}
+                          end,
+                      Parent ! {pmap_result, Result}
+                  end)
+        end,
+        L),
+    pmap_gather(length(L), []).
```
Comment (uwiger) [inline apps/aecore/src/aec_sync.erl:1213]: One way to write such a pmap (apologies for syntax errors):

```erlang
Pids = [spawn_monitor(fun() -> exit({ok, PerformTask(E)} end) || E <- L],
pmap_gather(Pids, []).

pmap_gather([], Acc) ->
    Acc;
pmap_gather([{Pid, MRef}|Pids], Acc) ->
    receive
        {'DOWN', MRef, process, Pid, Res} ->
            case Res of
                {ok, GoodRes} -> pmap_gather(Pids, [GoodRes | Acc]);
                Error -> do_something_clever()
            end
    end.
```

This also captures crashes, and makes it possible to tell the difference.

**Review by uwiger — COMMENTED** [review]
Questionmark regarding the apparently hard-coded numbers for verified/unverified peers. Otherwise LGTM.

File: apps/aecore/src/aec_peer_connection.erl (hunk 1 above)
Comment (velzevur) [reply apps/aecore/src/aec_peer_connection.erl:1475]: Oops

File: apps/aeutils/src/aeu_info.erl [hunk 4]
```diff
@@ -12,6 +13,24 @@ get_version() ->
...
 
+get_os() ->
+    Bin = fun(A) when is_atom(A) -> atom_to_binary(A, utf8);
+              (I) when is_integer(I) -> integer_to_binary(I)
+          end,
+    {OSFamily, OSName} = os:type(),
+    BOSFamily = Bin(OSFamily),
+    BOSName = Bin(OSName),
+    OSVsnBin =
+        case os:version() of
```
Comment (gorbak25) [inline apps/aeutils/src/aeu_info.erl:24]: Is the OS version an overkill?

File: apps/aecore/src/aec_peer_connection.erl [hunk 5]
```diff
@@ -1443,3 +1461,32 @@ validate_micro_signature(MicroHeader, _PrevHeader, PrevKeyHeader)
     end;
 validate_micro_signature(_MicroHeader, _PrevHeader, not_found) ->
     {error, signer_not_found}.
+
+%% -- Node info  message -----------------------------------------------------------
+handle_get_node_info(S) ->
+    case is_node_info_sharing_enabled() of
+        false -> S;
+        true ->
+            NodeVersion = integer_to_binary(?KEY_HEADER_INFO_LIMA_POINT_RELEASE),
```
Comment (gorbak25) [inline apps/aecore/src/aec_peer_connection.erl:1470]:
```
NodeVersion = aeu_info:get_version(),
NodeRevision = aeu_info:get_revision(),
```
The version alone should be ok.

File: apps/aecore/src/aec_sync.erl [hunk 6]
```diff
@@ -50,6 +52,15 @@ get_generation(PeerId, Hash) ->
 set_last_generation_in_sync() ->
     gen_server:cast(?MODULE, set_last_generation_in_sync).
 
+ask_all_for_node_info() ->
```
Comment (gorbak25) [inline apps/aecore/src/aec_sync.erl:55]: Should it be exposed by HTTP?

**Review by gorbak25 — CHANGES_REQUESTED** [review]
The version in the info msg can get out of sync with the real version.

File: apps/aecore/src/aec_sync.erl (hunk 6 above)
Comment (dincho) [reply apps/aecore/src/aec_sync.erl:55]: I wouldn't make this part of the node at all, it's a reporting functionality IMO.

Comment (velzevur) [reply apps/aecore/src/aec_sync.erl:55]: I'd rather keep it as an internal API which we call on demand. My concern would be an abuse of this API.

File: apps/aecore/src/aec_peer_connection.erl [hunk 7]
```diff
@@ -1443,3 +1461,34 @@ validate_micro_signature(MicroHeader, _PrevHeader, PrevKeyHeader)
...
+handle_get_node_info(S) ->
+    case is_node_info_sharing_enabled() of
+        false -> S;
+        true ->
+            NodeVersion = aeu_info:get_version(),
+            NodeRevision = aeu_info:get_revision(),
+            OS = aeu_info:get_os(),
+            Verified = aec_peers:count(verified),
+            Unverified = aec_peers:count(unverified),
+            NodeInfo = #{ version => NodeVersion
```
Comment (gorbak25) [inline apps/aecore/src/aec_peer_connection.erl:1475]: Revision might be an overkill, the vendor is lacking - what if alternate AE implementations pop up?
https://github.com/ethereum/devp2p/blob/master/rlpx.md#hello-0x00

Comment (velzevur) [reply apps/aecore/src/aec_peer_connection.erl:1475]: added a vendor

**Review by gorbak25 — DISMISSED** [review]

**Review by gorbak25 — APPROVED** [review]

**Review by uwiger — APPROVED** [review]

**Pull request merged**